onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -324,6 +330,121 @@ const std::vector<std::string>& GetAvailableProviders() {
   return available_providers;
 }
 
+#ifdef USE_CUDA
+
+/*
+ * Validate a string that is positive integer or zero.
+ *
+ * (-1234, 43.21, +43.21 ... are not valid,
+ *  1234, +4321, 12 ... are valid) 
+ */
+bool IsPositiveInteger(const std::string & s) {
+
+  if (s.length() == 0) {
+    return false;
+  }
+
+  for (size_t i = 0; i < s.length(); i++) {
+    if (i == 0 && s[i] == '+' && s.length() > 1) {
+      continue;
+    }
+
+    if (!isdigit(s[i])) {
+      return false; 
+    }      
+  }
+
+  return true;
+}
+
+bool IsCudaDeviceIdValid(InferenceSession* sess, int id) {
+    
+    int num_devices = 0;
+    CUDA_CALL_THROW(cudaGetDeviceCount(&num_devices));
+
+    if (0 == num_devices)
+    {
+        LOGS(*(sess->GetLogger()), WARNING) << "your system does not have a CUDA capable device.";
+        return false;
+    }
+
+    if (id < 0 || id >= num_devices)
+    {
+        LOGS(*(sess->GetLogger()), WARNING) << "cuda_device=" << id << " is invalid, must choose device ID between 0 and " << num_devices - 1;
+        return false;
+    }
+
+    return true;
+}
+
+
+void UpdateCudaProviderOptions(InferenceSession* sess, onnxruntime::CudaProviderOptions& options, std::unordered_map<std::string, std::string> options_map) {
+    std::unordered_map<std::string, std::string>::iterator it;
+
+    it = options_map.find("device_id");
+    if (it != options_map.end()) {
+        OrtDevice::DeviceId device_id;
+        try {
+          device_id = std::stoi(it->second); 
+        }
+        catch (const std::invalid_argument& ia) {
+            throw std::runtime_error("Please provide device id with integer.");
+        }
+
+        if (!IsCudaDeviceIdValid(sess, device_id)) {
+            throw std::runtime_error("Please provide available device id.");
+        }
+        options.device_id = device_id; 
+        LOGS(*(sess->GetLogger()), INFO) << "cuda device id is set to " << device_id;
+    }
+
+    it = options_map.find("cuda_mem_limit");
+    if (it != options_map.end()) {
+
+        // The reason to check whether the string is positive integer upfront is that
+        // when calling stoull(), if the minus sign was part of the input sequence, 
+        // the numeric value calculated from the sequence of digits is negated.
+        // In other words, it will cause wraparound. 
+        // So, we rule out negative integer string beforehand. 
+        if (!IsPositiveInteger(it->second)) {
+            throw std::runtime_error("Please provide cuda memory limitation size with positive integer.");
+        }
+
+        size_t size;
+        try {
+          size = std::stoull(it->second, nullptr, 0);
```

Comment:
https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

long long is 64 bits. but size_t could be 32 bits.

Reply:
I've changed to following:

#if (defined(__amd64__) || defined(_M_AMD64) || defined(__aarch64__))
    size = std::stoull(it->second, nullptr, 0);
#else
    size = std::stoul(it->second, nullptr, 0);
#endif

As from the link you provided, all the 64-bits data models use 64 bits to implement size_t and unsigned long long.
Therefore, they can be assigned to each other interchangebly. 
For 32-bits architecture, we use size_t and unsigned long since they both uses 32 bits.